NorthstarDLL/shared/gamepresence.cpp

```diff
@@ -0,0 +1,81 @@
+#include "pch.h"
+
+#include "gamepresence.h"
+#include "plugins/pluginbackend.h"
+#include "plugins/plugins.h"
+#include "dedicated/dedicated.h"
+#include "server/serverpresence.h"
+#include "masterserver/masterserver.h"
+#include "squirrel/squirrel.h"
+
+GameStatePresence* g_pGameStatePresence;
+
+GameStatePresence::GameStatePresence()
+{
+	g_pServerPresence->AddPresenceReporter(&m_GameStateServerPresenceReporter);
+}
+
+void GameStateServerPresenceReporter::RunFrame(double flCurrentTime, const ServerPresence* pServerPresence)
+{
+	g_pGameStatePresence->id = pServerPresence->m_sServerId;
+	g_pGameStatePresence->name = pServerPresence->m_sServerName;
+	g_pGameStatePresence->description = pServerPresence->m_sServerDesc;
+	g_pGameStatePresence->password = pServerPresence->m_Password;
+
+	g_pGameStatePresence->map = pServerPresence->m_MapName;
+	g_pGameStatePresence->playlist = pServerPresence->m_PlaylistName;
+	g_pGameStatePresence->current_players = pServerPresence->m_iPlayerCount;
+	g_pGameStatePresence->max_players = pServerPresence->m_iMaxPlayers;
+
+	g_pGameStatePresence->is_local = !IsDedicatedServer();
+}
+
+void GameStatePresence::RunFrame()
+{
+	if (g_pSquirrel<ScriptContext::UI>->m_pSQVM != nullptr && g_pSquirrel<ScriptContext::UI>->m_pSQVM->sqvm != nullptr)
+		g_pSquirrel<ScriptContext::UI>->Call("GenerateUIPresence");
```

Comment:
these squirrel functions should probably have less general names to indicate that they're called from code, similar to how respawn code names all their ones CodeCallback